File: plugin/pkg/telemetry/telemetry.go
// Package telemetry reads measured node usage. The refresh runs in the background, so a
// decision never calls the API; it reads the last sample and carries its age.
package telemetry

import (
	"context"
	"fmt"
	"sync"
	"time"

	v1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	restclient "k8s.io/client-go/rest"
	"k8s.io/klog/v2"
	metricsclient "k8s.io/metrics/pkg/client/clientset/versioned"
)

// Usage is one node's measured consumption, with the age of the measurement itself.
type Usage struct {
	MilliCPU    int64
	MemoryBytes int64
	AgeSeconds  float64
}

// sample dates the values from when the source says they were measured, not from the download.
type sample struct {
	milliCPU    int64
	memoryBytes int64
	measuredAt  time.Time
}

// Source is what the snapshot builder sees; nil means no telemetry, reported as missing.
type Source interface {
	Get(node string) (Usage, bool)
	Name() string
}

// Kubelet is reserved: it can be added without touching the strategies.
const (
	Disabled   = "off"
	MetricsAPI = "metrics-api"
	Kubelet    = "kubelet"
)

// Options configure a collector: refresh interval, and how old a measurement may be.
type Options struct {
	Config   *restclient.Config
	Interval time.Duration
	MaxAge   time.Duration
}

// Open returns the configured collector, or nil for no telemetry at all.
func Open(ctx context.Context, collector string, opts Options) (Source, error) {
	switch collector {
	case "", Disabled:
		return nil, nil
	case MetricsAPI:
		return newMetricsAPI(ctx, opts)
	case Kubelet:
		return nil, fmt.Errorf("telemetry collector %q is not implemented yet; it is reserved "+
			"for reading the kubelet summary endpoints directly, without metrics-server", collector)
	default:
		return nil, fmt.Errorf("unknown telemetry collector %q, known: %s, %s (%s reserved)",
			collector, Disabled, MetricsAPI, Kubelet)
	}
}

type client struct {
	api      metricsclient.Interface
	interval time.Duration
	maxAge   time.Duration

	mutex     sync.RWMutex
	samples   map[string]sample
	updatedAt time.Time
}

// newMetricsAPI starts a background refresher. A failing refresh leaves the previous sample
// in place until it ages out, so telemetry loss degrades to a fallback rather than to zeros.
func newMetricsAPI(ctx context.Context, opts Options) (Source, error) {
	api, err := metricsclient.NewForConfig(opts.Config)
	if err != nil {
		return nil, err
	}
	source := &client{api: api, interval: opts.Interval, maxAge: opts.MaxAge, samples: map[string]sample{}}
	if err := source.refresh(ctx); err != nil {
		klog.ErrorS(err, "kaptain telemetry: first refresh failed, starting without a sample",
			"collector", source.Name())
	}
	go source.loop(ctx)
	return source, nil
}

func (c *client) Name() string { return MetricsAPI }

func (c *client) loop(ctx context.Context) {
	ticker := time.NewTicker(c.interval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			if err := c.refresh(ctx); err != nil {
				klog.V(2).ErrorS(err, "kaptain telemetry: refresh failed")
			}
		}
	}
}

func (c *client) refresh(ctx context.Context) error {
	listed, err := c.api.MetricsV1beta1().NodeMetricses().List(ctx, metav1.ListOptions{})
	if err != nil {
		return err
	}
	samples := make(map[string]sample, len(listed.Items))
	for _, item := range listed.Items {
		cpu := item.Usage[v1.ResourceCPU]
		memory := item.Usage[v1.ResourceMemory]
		samples[item.Name] = sample{
			milliCPU:    cpu.MilliValue(),
			memoryBytes: memory.Value(),
			// metrics-server serves a cached measurement: dating it from the download here
			// would make a stale value look fresh.
			measuredAt: item.Timestamp.Time,
		}
	}
	c.mutex.Lock()
	defer c.mutex.Unlock()
	c.samples, c.updatedAt = samples, time.Now()
	return nil
}

// Get refuses a sample without a usable timestamp: an age that cannot be known cannot be
// checked against the limit.
func (c *client) Get(node string) (Usage, bool) {
	c.mutex.RLock()
	entry, ok := c.samples[node]
	c.mutex.RUnlock()
	if !ok || entry.measuredAt.IsZero() {
		return Usage{}, false
	}

	age := time.Since(entry.measuredAt)
	if age < 0 {
		age = 0 // clock skew between us and the metrics source
	}
	if age > c.maxAge {
		return Usage{}, false
	}
	return Usage{
		MilliCPU:    entry.milliCPU,
		MemoryBytes: entry.memoryBytes,
		AgeSeconds:  age.Seconds(),
	}, true
}

File: bridge/telemetry.py
"""Measured node usage, read from the Kubernetes metrics API.

Requested resources say what pods asked for, not what is being used. The BT arm of the
protocol needs the measured value, and so does any comparison claiming that a model beats
a heuristic *given the same observations*.

Same contract as the Go plugin's `pkg/telemetry`: a background refresh, a decision that
never calls the API, and an age carried with every sample so a stale value is visible.
"""

import logging
import threading
import time
from datetime import datetime, timezone

from snapshot import parse_cpu, parse_memory

log = logging.getLogger("kaptain.telemetry")

METRICS_GROUP = "metrics.k8s.io"
METRICS_VERSION = "v1beta1"


def parse_node_metrics(items: list[dict]) -> dict[str, dict]:
    """Node usage from the metrics API payload, in millicores and bytes.

    `measured_at` comes from the payload's own `timestamp`, not from the moment we
    downloaded it. metrics-server serves a cached sample, so a fresh fetch of a stale
    measurement would otherwise look fresh, and a strategy would rank on old numbers while
    reporting an age of milliseconds.
    """
    usage = {}
    for item in items:
        name = (item.get("metadata") or {}).get("name")
        if not name:
            continue
        measured = item.get("usage") or {}
        usage[name] = {
            "millicpu": parse_cpu(measured.get("cpu", "0")),
            "memory_bytes": parse_memory(measured.get("memory", "0")),
            "measured_at": parse_timestamp(item.get("timestamp")),
        }
    return usage


def parse_timestamp(value: str | None) -> float | None:
    """RFC 3339 timestamp to epoch seconds, or None when the API did not send one."""
    if not value:
        return None
    text = value.replace("Z", "+00:00")
    try:
        parsed = datetime.fromisoformat(text)
    except ValueError:
        return None
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    return parsed.timestamp()


class MetricsApiTelemetry:
    def __init__(self, refresh_seconds: float = 2.0, max_age_seconds: float = 30.0):
        self.refresh_seconds = refresh_seconds
        self.max_age_seconds = max_age_seconds
        self._lock = threading.Lock()
        self._usage: dict[str, dict] = {}
        self._updated_at: float | None = None
        self._api = None

    def start(self) -> None:
        from kubernetes import client, config

        config.load_incluster_config()
        self._api = client.CustomObjectsApi()
        try:
            self.refresh()
        except Exception as exc:
            log.error("telemetry: first refresh failed, starting without a sample: %s", exc)
        threading.Thread(target=self._loop, daemon=True, name="telemetry-refresh").start()

    def _loop(self) -> None:
        while True:
            time.sleep(self.refresh_seconds)
            try:
                self.refresh()
            except Exception as exc:
                log.warning("telemetry refresh failed: %s", exc)

    def refresh(self) -> None:
        listed = self._api.list_cluster_custom_object(METRICS_GROUP, METRICS_VERSION, "nodes")
        usage = parse_node_metrics(listed.get("items", []))
        with self._lock:
            self._usage = usage
            self._updated_at = time.monotonic()

    def get(self, node: str) -> dict | None:
        """The node's measured usage, or None when it is missing or too old.

        The age is the age of the measurement itself, per node. A sample without a usable
        timestamp is refused: an unknown age cannot be checked against the limit.
        """
        with self._lock:
            entry = self._usage.get(node)
        if entry is None or entry.get("measured_at") is None:
            return None
        age = time.time() - entry["measured_at"]
        if age < 0:
            age = 0.0          # clock skew between us and the metrics source
        if age > self.max_age_seconds:
            return None
        return {"millicpu": entry["millicpu"], "memory_bytes": entry["memory_bytes"],
                "age_seconds": age}
